### ISSUE TITLE
Endless loop in certain scenarios for ScalarEncoder's topDownCompute

### ISSUE THREAD

**Issue by clumsy** [numenta/htm.java#142]
Maybe it's me, but it seems like when I create a ScalarEncoder with, say, minVal = 0.0 it looks like one of the following happens:
The encoder is put into incorrect state right from the start with resolution set to 0 by default and not changed due to a check in initEncoder. Is 0.0 prohibited or is it put there to check for default values?
0.0 seems like a valid value to me, it's that they can't be both 0.0 with maxVal that would make sense to me. This results in an endless loop when calling topDownCompute as it iterates with 0.0 increment - that is forever.

It looks as if the invariants are not properly preserved, if I did something wrong during construction a RuntimeException should have been thrown when invoking ScalarEncoder.build() method.

The following test results in OutOfMemoryError due to endless append operation to an primitive collection:

```
@Test
	public void endlessLoopInTopDownCompute() {
		ScalarEncoder encoder = ScalarEncoder.builder()
    		.w( 5 )
    		.n( 10 )
    		.forced( true )
    		.minVal( 0 )
    		.maxVal( 100 )
    		.build();
		encoder.topDownCompute( new int[] { 0, 0, 0, 0, 0, 0, 0, 0, 0, 1 } );
	}
```

**Comment by cogmission** [comment]
@clumsy Thanks for pointing this out. I will take a look.

**Comment by cogmission** [comment]
Fixed by #185
